rfcs/0007-synchronous-commit.md

```diff
@@ -0,0 +1,446 @@
+
+    alt sync is Remote
+        WriteBatch->>WAL: await last applied position
+    end
+
+    MemTable-->>WriteBatch: wake up waiters of last applied position
+```
+
+### Handling No WAL Write
+
+SlateDB also support a no-WAL mode. In this mode, the write operation directly applied to MemTable without appending to the WAL. And the write operation can await the data to be persisted to storage as the specified durability level.
+
+In the new proposal, the read path need not to be changed, the readers could still read the WAL and Memtable as before.
+
+The write path needs some adjustments. The no-WAL mode will still maintain an in-memory WAL, but it's never flushed to storage.
+
+For `SyncLevel::Off`, the write operation could be considered as apply to MemTable immediately, and will be considered as committed.
+
+For `SyncLevel::Remote`, the write operation could be considered as append to an in-memory WAL, and directly flush them to L0 SST after the WAL buffer + MemTable reaches the memtable's L0 size threshold.
+
+The sequence diagram is like this:
+
+```mermaid
+sequenceDiagram
+    participant WriteBatch
+    participant WAL
+    participant MemTable
+    participant Flusher
+
+    WriteBatch->>WAL: append write op to current WAL
+    Note over WAL: become visible to readers with ReadWatermark::LastCommitting
+
+    WAL->>WAL: Maybe trigger a flush
+    Note over WAL: When the WAL buffer + MemTable reaches the memtable's L0 size threshold, trigger a flush.
+
+    WAL->>MemTable: Flush L0 SST
+    MemTable->>Flusher: Flush L0 SST
+
+    Flusher-->>MemTable: ack Flush L0 SST message, increment last L0 flushed position
+
+    MemTable->>MemTable: apply write op to MemTable, and freeze it into IMM
+    Note over MemTable: become visible to readers with ReadWaterMark::LastCommitted 
+
+    alt sync is Remote
+        WriteBatch->>WAL: await last applied position
+    end
+
+    MemTable-->>WriteBatch: wake up waiters of last applied position
+```
+
+### Handling WAL Write Failures
+
+It's more likely to have WAL write failures on object storage like S3 than local disk.
+
+We do not have to fail the write operation immediately when the WAL write fails. Instead, we can retry the write operation with a backoff.
+
+However, it's possible to have some failure on flushing WAL to object storage which last for several minutes or even longer, and without a local disk which might help to buffer the WAL during the failure.
+
+In this case, we can still buffer the WAL in memory as long as possible. When the failure is resolved, we can resume the flush operation.
+
+However, the memory is limited, and we can't buffer the WAL in memory forever. We need to set a threshold like named as `max_wal_mem_buffer_bytes` to set the max size of the WAL buffer if the flush operation always fails. When this threshold is reached, we can't buffer the WAL anymore, and we have to mark the db into an read-only state.
```

Comment:
Can we use `max_unflushed_bytes` for this?

```
    /// Defines the max number of unflushed key/value pair bytes that should reside in memory
    /// before applying backpressure to writers. This includes key/value pairs in both the
    /// immutable WAL flush queue and the immutable memtable flush queue. Writes will be
    /// paused if the total number of unflushed bytes exceeds this value until data is flushed
    /// to object storage.
    pub max_unflushed_bytes: usize,
```

(in config.rs)

Reply:
wow, the semantics is exactly aligned, thank you for sharing this!